Include/internal/pycore_critical_section.h

```diff
@@ -0,0 +1,204 @@
+#ifndef Py_INTERNAL_CRITICAL_SECTION_H
+#define Py_INTERNAL_CRITICAL_SECTION_H
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include "pycore_lock.h"        // PyMutex
+#include "pycore_pystate.h"     // _PyThreadState_GET()
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Implementation of Python critical sections: helpers to replace the global
+// interpreter lock with fine grained locking.
+//
+// A Python critical section is a region of code that can only be executed by
+// a single thread at at time. The regions begin with a call to
+// Py_BEGIN_CRITICAL_SECTION and ends with Py_END_CRITICAL_SECTION. Active
+// critical sections are *implicitly* suspended whenever the thread calls
+// `_PyThreadState_Detach()` (i.e., at tiems when the interpreter would have
+// released the GIL).
+//
+// The most recent critical section is resumed when `_PyThreadState_Attach()`
+// is called. See `_PyCriticalSection_Resume()`.
+//
+// The purpose of implicitly ending critical sections is to avoid deadlocks
+// when locking multiple objects. Any time a thread would have released the
+// GIL, it releases all locks from active critical sections. This includes
+// times when a thread blocks while attempting to acquire a lock.
+//
+// Note that the macros Py_BEGIN_CRITICAL_SECTION and Py_END_CRITICAL_SECTION
+// are no-ops in non-free-threaded builds.
+//
+// Example usage:
+//  Py_BEGIN_CRITICAL_SECTION(op);
+//  ...
+//  Py_END_CRITICAL_SECTION;
+//
+// To lock two objects at once:
+//  Py_BEGIN_CRITICAL_SECTION2(op1, op2);
+//  ...
+//  Py_END_CRITICAL_SECTION2;
+
+
+// Tagged pointers to critical sections use the two least significant bits to
+// mark if the pointed-to critical section is inactive and whether it is a
+// _PyCriticalSection2 object.
+#define _Py_CRITICAL_SECTION_INACTIVE       0x1
+#define _Py_CRITICAL_SECTION_TWO_MUTEXES    0x2
+#define _Py_CRITICAL_SECTION_MASK           0x3
+
+#ifdef Py_NOGIL
+#define Py_BEGIN_CRITICAL_SECTION(op) {                             \
+    _PyCriticalSection _cs;                                          \
+    _PyCriticalSection_Begin(&_cs, &_PyObject_CAST(op)->ob_mutex)
+
+#define Py_END_CRITICAL_SECTION                                     \
```

Comment:
FWIW, the mismatch with `Py_BEGIN_CRITICAL_SECTION` (i.e. args vs. no args) is a bit of an itch for me.  Consider adding empty args:

```suggestion
#define Py_END_CRITICAL_SECTION()                                   \
```

Reply:
I've updated it to use `()`.